presto-main/src/main/java/com/facebook/presto/operator/aggregation/AreaUnderROCCurveAggregation.java

```diff
@@ -0,0 +1,217 @@
+    {
+        MultiKeyValuePairs pairs = state.get();
+        if (pairs == null) {
+            pairs = new MultiKeyValuePairs(state.getKeyType(), state.getValueType());
+            state.set(pairs);
+        }
+
+        long startSize = pairs.estimatedInMemorySize();
+        pairs.add(label, score, position, position);
+        state.addMemoryUsage(pairs.estimatedInMemorySize() - startSize);
+    }
+
+    public static void combine(MultiKeyValuePairsState state, MultiKeyValuePairsState otherState)
+    {
+        if (state.get() != null && otherState.get() != null) {
+            Block labelsBlock = otherState.get().getKeys();
+            Block scoresBlock = otherState.get().getValues();
+            MultiKeyValuePairs pairs = state.get();
+            long startSize = pairs.estimatedInMemorySize();
+            for (int i = 0; i < labelsBlock.getPositionCount(); i++) {
+                pairs.add(labelsBlock, scoresBlock, i, i);
+            }
+            state.addMemoryUsage(pairs.estimatedInMemorySize() - startSize);
+        }
+        else if (state.get() == null) {
+            state.set(otherState.get());
+        }
+    }
+
+    public static void output(MultiKeyValuePairsState state, BlockBuilder out)
+    {
+        MultiKeyValuePairs pairs = state.get();
+        if (pairs == null) {
+            out.appendNull();
+        }
+        else {
+            Block labelsBlock = pairs.getKeys();
+            Block scoresBlock = pairs.getValues();
+            boolean[] labels = new boolean[labelsBlock.getPositionCount()];
+            double[] scores = new double[labelsBlock.getPositionCount()];
+            for (int i = 0; i < labelsBlock.getPositionCount(); i++) {
+                labels[i] = BOOLEAN.getBoolean(labelsBlock, i);
+                scores[i] = DOUBLE.getDouble(scoresBlock, i);
+            }
+            double auc = computeAUC(labels, scores);
+            if (Double.isFinite(auc)) {
+                DOUBLE.writeDouble(out, auc);
+            }
+            else {
+                out.appendNull();
+            }
+        }
+    }
+
+    /**
+     * Calculate Area Under the ROC Curve using trapezoidal rule
+     * See Algorithm 4 in a paper "ROC Graphs: Notes and Practical Considerations
+     * for Data Mining Researchers" for detail:
+     * http://www.hpl.hp.com/techreports/2003/HPL-2003-4.pdf
+     */
+    private static double computeAUC(boolean[] labels, double[] scores)
+    {
+        // labels sorted descending by scores
+        Integer[] sortedIndices = IntStream.range(0, labels.length).boxed().sorted((i, j) -> (int) Math.signum(scores[j] - scores[i])).toArray(Integer[]::new);
+
+        int tp = 0;
+        int fp = 0;
+
+        int tpPrev = 0;
+        int fpPrev = 0;
+
+        double area = 0.0;
+
+        double scorePrev = Double.POSITIVE_INFINITY;
+
+        for (int i : sortedIndices) {
+            if (scores[i] != scorePrev) {
```

Comment:
you are doing a double equality check here, better to check `Double.compare(d1, d2) == 0`